log/keyvalue_test.go

```diff
@@ -0,0 +1,302 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package log_test
+
+import (
+	golog "log"
+	"os"
+	"testing"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/logr/testr"
+	"github.com/go-logr/stdr"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/otel/internal/global"
+	"go.opentelemetry.io/otel/log"
+)
+
+func TestKind(t *testing.T) {
+	testCases := []struct {
+		kind  log.Kind
+		str   string
+		value int
+	}{
+		{log.KindBool, "Bool", 1},
+		{log.KindBytes, "Bytes", 5},
+		{log.KindEmpty, "Empty", 0},
+		{log.KindFloat64, "Float64", 2},
+		{log.KindInt64, "Int64", 3},
+		{log.KindSlice, "Slice", 6},
+		{log.KindMap, "Map", 7},
+		{log.KindString, "String", 4},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.str, func(t *testing.T) {
+			assert.Equal(t, tc.value, int(tc.kind), "Kind value")
+			assert.Equal(t, tc.str, tc.kind.String(), "Kind string")
+		})
+	}
+}
+
+func TestValueEqual(t *testing.T) {
+	vals := []log.Value{
+		{},
+		log.Int64Value(1),
+		log.Int64Value(2),
+		log.Float64Value(3.5),
+		log.Float64Value(3.7),
+		log.BoolValue(true),
+		log.BoolValue(false),
+		log.StringValue("hi"),
+		log.StringValue("bye"),
+		log.BytesValue([]byte{1, 3, 5}),
+		log.SliceValue(log.StringValue("foo")),
+		log.SliceValue(log.IntValue(3), log.StringValue("foo")),
+		log.MapValue(log.Bool("b", true), log.Int("i", 3)),
+		log.MapValue(
+			log.Slice("l", log.IntValue(3), log.StringValue("foo")),
+			log.Bytes("b", []byte{3, 5, 7}),
+		),
+	}
+	for i, v1 := range vals {
+		for j, v2 := range vals {
+			assert.Equal(t, i == j, v1.Equal(v2), "%v.Equal(%v)", v1, v2)
```

Comment:
The failure messages from assertions like this one (I changed the test to fail) convinced me to implement `Stringer`.

```
        	Error Trace:	/home/rpajak/repos/opentelemetry-go/log/keyvalue_test.go:81
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestValueEqual
        	Messages:   	{[] 2 0xc0001a4410}.Equal({[] 2 0xc0001a4410})
```

Reference: https://github.com/open-telemetry/opentelemetry-go/pull/4809#discussion_r1482157185

I suggest simply creating a separate issue for now.

Reply:
I created https://github.com/open-telemetry/opentelemetry-go/issues/4952